src/app/clusters/thermostat-server/thermostat-server.cpp

```diff
@@ -220,6 +220,293 @@ int16_t EnforceCoolingSetpointLimits(int16_t CoolingSetpoint, EndpointId endpoin
+
+    if (AbsMaxHeatSetpointLimit::Get(endpoint, &setpointLimits.absMaxHeatSetpointLimit) != Status::Success)
+        setpointLimits.absMaxHeatSetpointLimit = kDefaultAbsMaxHeatSetpointLimit;
+
+    if (MinHeatSetpointLimit::Get(endpoint, &setpointLimits.minHeatSetpointLimit) != Status::Success)
+        setpointLimits.minHeatSetpointLimit = setpointLimits.absMinHeatSetpointLimit;
+
+    if (MaxHeatSetpointLimit::Get(endpoint, &setpointLimits.maxHeatSetpointLimit) != Status::Success)
+        setpointLimits.maxHeatSetpointLimit = setpointLimits.absMaxHeatSetpointLimit;
+
+    if (setpointLimits.coolSupported)
+    {
+        if (OccupiedCoolingSetpoint::Get(endpoint, &setpointLimits.occupiedCoolingSetpoint) != Status::Success)
+        {
+            ChipLogError(Zcl, "Error: Can not read Occupied Cooling Setpoint");
+            return Status::Failure;
+        }
+    }
+
+    if (setpointLimits.heatSupported)
+    {
+        if (OccupiedHeatingSetpoint::Get(endpoint, &setpointLimits.occupiedHeatingSetpoint) != Status::Success)
+        {
+            ChipLogError(Zcl, "Error: Can not read Occupied Heating Setpoint");
+            return Status::Failure;
+        }
+    }
+
+    if (setpointLimits.coolSupported && setpointLimits.occupancySupported)
+    {
+        if (UnoccupiedCoolingSetpoint::Get(endpoint, &setpointLimits.unoccupiedCoolingSetpoint) != Status::Success)
+        {
+            ChipLogError(Zcl, "Error: Can not read Unoccupied Cooling Setpoint");
+            return Status::Failure;
+        }
+    }
+
+    if (setpointLimits.heatSupported && setpointLimits.occupancySupported)
+    {
+        if (UnoccupiedHeatingSetpoint::Get(endpoint, &setpointLimits.unoccupiedHeatingSetpoint) != Status::Success)
+        {
+            ChipLogError(Zcl, "Error: Can not read Unoccupied Heating Setpoint");
+            return Status::Failure;
+        }
+    }
+    return Status::Success;
+}
+
+/**
+ * @brief Checks to see if it's possible to adjust the heating setpoint to preserve a given deadband
+ * if the cooling setpoint is changed
+ *
+ * @param autoSupported Whether or not the thermostat supports Auto mode
+ * @param newCoolingSetpoing The desired cooling setpoint
+ * @param minHeatingSetpoint The minimum allowed heating setpoint
+ * @param deadband The deadband to preserve
+ * @return Success if the deadband can be preserved, InvalidValue if it cannot
+ */
+Status CheckHeatingSetpointDeadband(bool autoSupported, int16_t newCoolingSetpoint, int16_t minHeatingSetpoint, int16_t deadband)
+{
+    if (!autoSupported)
+    {
+        return Status::Success;
+    }
+    int16_t maxValidHeatingSetpoint = static_cast<int16_t>(newCoolingSetpoint - deadband);
+    if (maxValidHeatingSetpoint < minHeatingSetpoint)
+    {
+        // If we need to adjust the heating setpoint to preserve the deadband, it will go below the min heat setpoint
+        return Status::InvalidValue;
```

Comment:
So where does the spec say to return InvalidValue in this case?  Here's what I see in the spec, for example for the OccupiedCoolingSetpoint attribute:

> If an attempt is made to set this attribute to a value greater than MaxCoolSetpointLimit or less than MinCoolSetpointLimit, a response with the status code CONSTRAINT_ERROR SHALL be returned.
>
> If this attribute is set to a value that is less than (OccupiedHeatingSetpoint + MinSetpointDeadBand), the value of OccupiedHeatingSetpoint SHALL be adjusted to (OccupiedCoolingSetpoint - MinSetpointDeadBand).

That second sentence does not seem to envision what happens when the value of OccupiedHeatingSetpoint cannot thus be adjusted.  But that first items seems to suggest that trying to do the adjustment should result in ConstraintError, no?

Reply:
InvalidValue is the old, deprecated name...  But OK, maybe a followup to switch it globally here?

Reply:
OK, will leave for a follow-up PR.